doc: testing lambdas (#2340)
FusionAuth/fusionauth-site: site/docs/v1/tech/lambdas/testing-lambdas.md

> ---

Reviewer: so it should live under https://fusionauth.io/docs/v1/tech/guides/

title: Testing Lambdas With FusionAuth
description: Testing lambdas with FusionAuth
navcategory: customization
---

## Overview

This guide shows you how to create a simple lambda manually, update it programmatically, and test it with unit and integration tests. You can familiarize yourself with lambdas by reading the [FusionAuth lambda documentation](/docs/v1/tech/lambdas).

## Prerequisites

To follow this guide, you need

- [Node.js version 18](https://nodejs.org/en/download) or later, and
- [Docker](https://www.docker.com/get-started/).

### Lambda Limitations

{% include _callout-note.liquid content="<p>Remember the following limitations of lambdas when planning what they'll do:<ul><li>Lambdas do not have full access to JavaScript libraries, nor can they load them currently.</li> <li>The console methods take only one argument.</li><li>HTTP requests are not available in the Community or Starter FusionAuth plans.</li><li>If you set the Identity Provider <a href='/docs/v1/tech/identity-providers/#linking-strategies'>linking strategy</a> to 'Link Anonymously', no lambdas will be used for external authentication.</li> </ul></p>"%}

## Set Up The FusionAuth Sample Project

> Download or use Git to clone the [testing-lambdas repository](https://github.com/FusionAuth/fusionauth-example-testing-lambdas). Open a terminal in the directory you just created and start FusionAuth with Docker.

Reviewer: I moved that code over to a new repo, thanks!

```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCIsImd0eSI6WyJhdXRob3JpemF0aW9uX2NvZGUiXSwia2lkIjoiMWU1NmM0OWU4In0.eyJhdWQiOiJkZGQwNTAyMS0wNjgyLTQ4NWUtYThlMi1kMDMyOTY0YjAyMTEiLCJleHAiOjE2ODkyNjQwNzEsImlhdCI6MTY4OTI2MDQ3MSwiaXNzIjoiYWNtZS5jb20iLCJzdWIiOiIwYTkzOTYwNi0zNmVjLTQ1M2ItOTM0Mi04ZWZmOTE3ZjJhZWYiLCJqdGkiOiIyYmZlMjUwNy1hZWM0LTRjOTEtYWY5Yy1hOWVhYjQzNmQ4MGYiLCJhdXRoZW50aWNhdGlvblR5cGUiOiJQQVNTV09SRCIsImVtYWlsIjoiZXJsaWNoQGV4YW1wbGUuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImFwcGxpY2F0aW9uSWQiOiJkZGQwNTAyMS0wNjgyLTQ4NWUtYThlMi1kMDMyOTY0YjAyMTEiLCJzY29wZSI6Im9mZmxpbmVfYWNjZXNzIiwicm9sZXMiOltdLCJzaWQiOiIzNDk5MTAxMS1kNzUxLTRlOTctYWZiNi0zNzQ2N2RlYTc5YWIiLCJhdXRoX3RpbWUiOjE2ODkyNjA0NzEsInRpZCI6ImNiY2VkOWVhLWQ3NzgtZDBlYi03ZjU4LWE0MGYxY2VlNWFhYiIsIm1lc3NhZ2UiOiJIZWxsbyBXb3JsZCEifQ.3DOvP8LRAp6pIh0guUjJjYbNwZKzruVWre8Xq8x_S8k
```

> Copy this token from your terminal and paste it into the <span>Token</span>{:.field} text box on the [FusionAuth Online JWT Decoder](https://fusionauth.io/dev-tools/jwt-decoder). You'll see `"message": "Hello World!"` in the <span>Payload</span>{:.uielement} box, showing you that your new lambda ran correctly.

Reviewer: ```suggestion
Copy this token from your terminal and paste it into the <span>Token</span>{:.field} text box on the [FusionAuth Online JWT Decoder](/dev-tools/jwt-decoder). You'll see `"message": "Hello World!"` in the <span>Payload</span>{:.uielement} box, showing you that your new lambda ran correctly.
```

Please remove all absolute URL references to fusionauth.io.

## Programmatically Update A Lambda

Let's take a look at how to update your lambda programmatically using the FusionAuth API. Refer to the [APIs](/docs/v1/tech/apis/) documentation for more.

### Understand The Client Libraries

Although you can use the [lambda API](/docs/v1/tech/apis/lambdas) directly by making HTTP requests, it's much easier to use one of the provided [client libraries](/docs/v1/tech/client-libraries/).

> There are two libraries for Javascript:

Reviewer: ```suggestion
There are two ways to do this using Javascript:
```

- The [TypeScript client library](https://github.com/FusionAuth/fusionauth-typescript-client), documented [here](/docs/v1/tech/client-libraries/typescript), should be used for any browser or Node code you write in JavaScript or TypeScript. It provides a straightforward way of calling the underlying HTTP API.

> - The [Node CLI](https://github.com/FusionAuth/fusionauth-node-cli) is a set of commands you can run in the terminal to perform a few advanced functions, such as uploading a new theme or lambda to your FusionAuth application. The Node CLI is a wrapper on the TypeScript client library and operates at a higher level of abstraction. It is helpful to manage lambdas, but you can always drop down to the Typescript client library if needed.

Reviewer: ```suggestion
- The [Node CLI](https://github.com/FusionAuth/fusionauth-node-cli) is a set of commands you can run in the terminal to perform a certain functions. The focus of this CLI is on uploading and downloading of commonly modified assets such as lambdas or themes. The Node CLI is a wrapper on the TypeScript client library and operates at a higher level of abstraction. It is helpful to manage lambdas, but you can always drop down to the Typescript client library if needed.
```

### Create An API Key

The API, CLI, and Client Library all need an API Key to access FusionAuth.

> The kickstart configuration file used by FusionAuth already created a sample API Key with superuser privileges. For more information on managing API keys, please refer to the following [guide](https://fusionauth.io/docs/v1/tech/apis/authentication#managing-api-keys). If you need to create another key in future applications, you can perform the following steps but you don't need to for this tutorial.

Reviewer: This is confusing to me as a reader. Why are you showing me how to create another API key if there's already one that I can use? I don't understand.

I'd say either walk the user through this (in the interests of building a key with minimal permissions, which is a good practice) or delete this and use the superuser key (and of course, warn they shouldn'd do so in prod).

```bash
npx fusionauth lambda:retrieve f3b3b547-7754-452d-8729-21b50d111505 --key lambda_testing_key
```

The lambda will be saved to a file, where the file name is the UUID of your lambda. So it should look like this: `./lambdas/f3b3b547-7754-452d-8729-21b50d111505.yaml`.

Let's update the lambda to say "Goodbye World!" instead of "Hello World!" and re-upload it. Open the file in a text editor, change the value of the <span>body</span>{:.field} property to the following.

{% raw %}
```
body: |
  function populate(jwt, user, registration) {
    jwt.message = 'Goodbye World!';
    console.info('Goodbye World!');
  }
```
{% endraw %}

Save the file and upload the lambda with the following command.

```bash
npx fusionauth lambda:update f3b3b547-7754-452d-8729-21b50d111505 --key lambda_testing_key
```

You can check that the lambda in FusionAuth now says "Goodbye World!" by viewing the ["[ATest]" lambda details](http://localhost:9011/admin/lambda).

{% include docs/_image.liquid src="/assets/img/docs/customization/lambdas/testing-lambdas/updated-lambda.png" alt="Update lambda" class="img-fluid bottom-cropped" width="1200" figure=false %}

### Other CLI Functionality

> You can create a lambda using the Node CLI with `lambda:create` and delete a lambda with `lambda:delete`, but only if the lambda is not linked to an application. To link a lambda to an application, use `lambda:link-to-application` with existing application and lambda Ids. Linking a lambda will use it for both the application's <span>Access Token populate lambda</span>{:.field} and <span>Id Token populate lambda</span>{:.field}. To unlink a lambda from an application, use `lambda:unlink-from-application` with existing application and lambda Ids. Unlinking a lambda will remove it from both the fields or just one of them if the other links to a different lambda. To link types of lambdas other than JWT populate, use the underlying client library or the admin UI.

Reviewer: As discussed, let's replace this with a screenshot showing how to associate the lambda with an application. You can mention that this can be done via the client library or API as well.

#### Create A User

Before you can write any tests, you need a test user profile to log in with. The test app `package.json` includes a reference to the `@fusionauth/typescript-client` discussed earlier, so you can use that to create a test user programmatically.

Make a new file called `test.js` in the app and paste the following code into it.

```js
{% remote_include 'https://raw.githubusercontent.com/FusionAuth/fusionauth-example-testing-lambdas/main/complete-application/documentation_snippets/test_1.js' %}
```

The code above has three sections:

- A declaration of constant variables that match the `kickstart.json` and `.env` files.
- The `createRandomUser` function, which creates a User request object with `const request` and then sends it to `fusion.register()`. Details on this can be found in the [TypeScript client library interface](https://github.com/FusionAuth/fusionauth-typescript-client/blob/master/src/FusionAuthClient.ts).
- The `deleteUser` function, which you can use in the tests to delete the user just created.

Run the code to test user creation with the following command.

```bash
node test.js
```

> In FusionAuth, click on <span>Users</span>{:.breadcrumb} to check that a new user called `lambdatestuser` has been created. You can delete the `createRandomUser(uuidv4());` line in `test.js` as each test will use a new temporary user. This will allow you to add multiple lambda tests while avoiding potential conflicts between test users and permissions.

Reviewer: why did we call `createRandomUser` here just to delete it? Is this proving that the typescript library is configured correctly or for some other reason?

#### Write The Test

Now you will test that the lambda returns "Goodbye World", which will confirm that the CLI `update` command worked.

> Add this code to the `test.js` file.

Reviewer: Can you have multiple test files using the NPM test library? Looks like it: `tape tests/**/*.js` 

Then you could have each file live for real separately in the repo, rather than have users keep adding to `test.js`

Run it with the following command.

```bash
node test.js
```

When your code has several tests, and you want a colorful, concise summary, you can use the following instead.

```bash
node test.js | npx faucet
```

### Unit Test: Call An External Service

The next test you'll write is a unit test that verifies your lambda locally using a fake mock service and not in FusionAuth. The benefit of this test is that you can test your logic works without needing an external service to be reliable at the time of testing. The danger is that your test might pass locally, but the lambda might fail on FusionAuth due to it running on a different JavaScript environment with different restrictions and configuration.

Let's take an example where you check if users have email addresses from a country sanctioned by the United States, such as North Korea or Cuba. You call the external site `https://issanctioned.example.com` with an email address, and you're told whether or not the domain is banned.

Add this new function to `test.js`. 

```js

> {% remote_include 'https://raw.githubusercontent.com/FusionAuth/fusionauth-example-testing-lambdas/main/complete-application/documentation_snippets/test_3.js' %}

Reviewer: I don't understand the point of putting this into a documentation snippet? The point of pulling this from a remote server is that the code will be runnable and stay up to date. See for example the doc here: https://fusionauth.io/docs/v1/tech/themes/client-side-password-rule-validation

How does putting these into a documentation snippet help with this goal?

Reviewer: How about instead you build up a test suite that the user can run?

```

This is the lambda function that would run on FusionAuth, similar to the "Hello World" function described earlier.

Now add the two tests below. The first test checks if North Korea (`.kp`) is banned and the second if Canada (`.ca`) is allowed.

```js
{% remote_include 'https://raw.githubusercontent.com/FusionAuth/fusionauth-example-testing-lambdas/main/complete-application/documentation_snippets/test_4.js' %}
```

This test function uses `fetchMock` to mock the external service that would be called from the lambda function in FusionAuth. The mocks for the JWT, user, and registration objects are all simple `{}` objects you can pass as parameters to the `populate()` lambda.

Finally, run the tests.

```bash
node test.js  | npx faucet

> ```

Reviewer: Can we see sample output of the tests (all three)? No need for a screenshot, just copying and pasting the output is fine.

If your HTTP Connect fetch request fails when deployed to FusionAuth, please review the [documentation](/docs/v1/tech/lambdas/#using-lambda-http-connect). In particular, ensure you are using a license and have purchased the correct plan (Essentials or Enterprise).

### Unit Test: Populate JWT From FusionAuth

In this final unit test, let's look at how to check user information available in FusionAuth to determine custom fields to return to your app.

There are two objects to consider. The first is the JWT fields that are returned to your app by default when a user logs in.

```js
{
  aud: 'e9fdb985-9173-4e01-9d73-ac2d60d1dc8e',
  exp: 1692273965,
  iat: 1692270365,
  iss: 'acme.com',
  sub: '00000000-0000-0000-0000-111111111111',
  jti: '47cc65ce-3981-4e40-8c99-5221d69c53da',
  authenticationType: 'PASSWORD',
  email: 'richard@example.com',
  email_verified: true,
  applicationId: 'e9fdb985-9173-4e01-9d73-ac2d60d1dc8e',
  roles: [ 'Petty cash keeper', 'First aider' ],
  auth_time: 1692270365,
  tid: 'd7d09513-a3f5-401c-9685-34ab6c552453',
  message: 'goodbye world'
}
```

The second object is the user supplied to your `populate()` function in a lambda.

```js

> user: {

Reviewer: Please pull these from https://github.com/FusionAuth/fusionauth-site/tree/master/site/docs/src/json as those json files will be updated over time.

    active: true,
    birthDate: '1985-11-23',
    connectorId: 'e3306678-a53a-4964-9040-1c96f36dda72',
    data: {},
    email: 'richard@example.com',
    firstName: 'Richard',
    id: '00000000-0000-0000-0000-111111111111',
    insertInstant: 1692176681681,
    lastLoginInstant: 1692270415045,
    lastName: 'Hendricks',
    lastUpdateInstant: 1692176681681,
    memberships: [],
    passwordChangeRequired: false,
    passwordLastUpdateInstant: 1692176681694,
    preferredLanguages: [],
    registrations: [
      {
        applicationId: 'e9fdb985-9173-4e01-9d73-ac2d60d1dc8e',
        data: {},
        id: '631b3ef1-4f46-42d2-b315-8d29c1cf18ff',
        insertInstant: 1692176681698,
        lastLoginInstant: 1692270415045,
        lastUpdateInstant: 1692269815209,
        preferredLanguages: [],
        roles: [ 'Petty cash keeper', 'First aider' ],
        tokens: {},
        usernameStatus: 'ACTIVE',
        verified: true
      }
    ],
    tenantId: 'd7d09513-a3f5-401c-9685-34ab6c552453',
    twoFactor: { methods: [], recoveryCodes: [] },
    usernameStatus: 'ACTIVE',
    verified: true
  }
```

> You can see that the user object has data that the JWT does not, like names, birthdates, and languages, that you might want to add in a lambda.

Reviewer: I'm sorry I was vague here. What I'd love to do is see someone able to:

* download their lambda via the node CLI
* run a unit test against that lambda where the test code is outside of the directory, but the downloaded lambda is being tested
* have the user and other objects mocked up by the test

Does that make sense? What you provided is helpful, but the user will still have to copy and paste their lambda code into `test.js` every time it is downloaded.

Can we import the lambda function into the test context?

  - [Lambda Limitations](#lambda-limitations)
- [Manually Creating a Simple Lambda](#manually-creating-a-simple-lambda) 
- [Programmatically Updating a Lambda](#programmatically-updating-a-lambda) 
  - [Understanding the Client Libraries](#understanding-the-client-libraries) 
  - [Creating an API Key](#creating-an-api-key) 
  - [Using the Lambda CLI](#using-the-lambda-cli) 
  - [API limitations](#api-limitations) 
- [Testing Overview](#testing-overview)
  - [Test Library](#test-library) 
  - [Integration Test: Verify JWT population](#integration-test-verify-jwt-population) 
  - [Unit Test: Calling an External Service](#unit-test-calling-an-external-service)
- [Addendum - test.js](#addendum---testjs)

## Prerequisites

To follow this guide, you will need a test instance of FusionAuth. For FusionAuth installation instructions, please visit [the 5-minute setup guide](/docs/v1/tech/5-minute-setup-guide). 

This guide assumes FusionAuth can be accessed at `http://localhost:9011/admin` with your [FusionAuth example test app](/docs/v1/tech/getting-started/5-minute-docker#5-configure-the-backend-to-complete-the-login) at `http://localhost:3000`.

### Lambda Limitations

> {% include _callout-note.liquid content="<p>Remember the following limitations of lambdas when planning what they'll do:<ul><li>Lambdas do not have full access to JavaScript libraries, nor can they load them currently.</li> <li>The console methods take only one argument.</li><li>HTTP requests are not available in the Community or Starter FusionAuth plans.</li><li>If you set the Identity Provider <a href='/docs/v1/tech/identity-providers/#linking-strategies'>linking strategy</a> to 'Link Anonymously', no lambdas will be used for external authentication.</li> </ul></p>"%}

Reviewer: Can you break this out to a separate include file?

Would like to include them in the known limitations section as well.

Reviewer 2: added the file `site/_includes/docs/_lambda-limitations.adoc`

## Manually Creating a Simple Lambda

Let's start by making a simple lambda to test that it works on your machine.

- Log in to [FusionAuth admin](`http://localhost:9011/admin).
- <p>Navigate to <span class="breadcrumb">Customizations -> Lambdas</span>.</p>
- Click the <i/>{:.ui-button .green .fa .fa-plus} button at the top right to add a new lambda.
- <p>Leave the <span class="field">Id</span> blank.</p>
- <p>Enter the <span class="field">Name</span> "[ATest]" (to put it at the top of the list of lambdas alphabetically).</p>
- <p>Select the <span class="field">Type</span> "JWT Populate".</p>

> - <p>Leave the <span class="field">Engine</span> on "GraalJS".</p>

Reviewer: ```suggestion
- <p>Leave the <span class="field">Engine</span> as "GraalJS".</p>
```

- Click on the <span>JWT</span>{:.breadcrumb} tab.
- <p>Toggle <span class="uielement">Enabled</span> to on.</p>
- <p>Under <span class="breadcrumb">Lambda Settings</span>, select the lambda you created, called "[ATest]" for the <span class="field">Access Token populate lambda</span>.</p>
- Click the <i/>{:.ui-button .blue .fa .fa-save} button to save the changes.

{% include docs/_image.liquid src="/assets/img/docs/customization/lambdas/testing-lambdas/edit-application.png" alt="Enabling the lambda in an application" class="img-fluid" width="1200" figure=false %}

You can now test that the new lambda writes to the event log and returns extra data in the JWT. You need to open the `routes/index.js` file in the [FusionAuth example 5-minute guide test app code](https://github.com/FusionAuth/fusionauth-example-5-minute-guide) that you cloned from GitHub.

Add a line in the OAuth redirect route handler `routes/index.js` to write the JWT to the console at login. The start of the function should now look like the code below.

```js
router.get('/oauth-redirect', function (req, res, next) {

>   console.dir(res);

Author: @mooreds - this is just a standard JS, not part of Lambda -- like `console.log` but it shows all the properties of the object 

https://developer.mozilla.org/en-US/docs/Web/API/console/dir

should we add a note about this?

>   console.dir(res);

Reviewer: No, if it is something a JS developer would know, no reason to add a note. Thanks!

- The [Node CLI](https://github.com/FusionAuth/fusionauth-node-cli) is a set of commands you can run in the terminal to perform a few advanced functions, such as uploading a new theme or lambda to your FusionAuth application. The Node CLI is a wrapper on the TypeScript client library. This is a small library you'll use only in a few cases.

### Creating an API Key

Before you can use any of the API, CLI, or Client Library functionality, you need to create an API Key to allow you to access FusionAuth.

- Navigate to  <span>Settings -> API Keys</span>{:.breadcrumb} and click the <i/>{:.ui-button .green .fa .fa-plus} button at the top right.
- <p>In the <span class="field">Key</span> field enter "lambda_testing_key".</p>
- You can either leave all the toggle buttons for the Endpoints blank to give this key all permissions, or enable the buttons only for the following endpoints:
  - `/api/lambda`
  - `/api/lambda/search`
  - `/api/login`
  - `/api/user`
  - `/api/user/registration`
- Save the API key.

### Using the Lambda CLI

First, install the Node CLI library. Open a terminal in your app folder and use the following commands.

```bash

> npm install --save-dev @fusionauth/cli;

Reviewer 2: Local npm package installation is safer than global, and does not usually cause version conflicts

      throw Error(clientResponse);
    const jwtToken = clientResponse.response.token;
    const decodedToken = jwt.decode(jwtToken);
    const message = decodedToken.message;
    return message;
  }
  catch (e) {
    console.error('Error: ');
    console.dir(e, { depth: null });
    process.exit(1);
  }
}
```

The `login` function calls the FusionAuth library in a similar way to `createUser`. It then decodes the JWT response and returns its `message` property.

This `login` function is called by the `tape` function `test`. This test gives the name of the test, says that it expects exactly one assertion to occur with `plan`, checks that calling `login` returns the property you expect from our lambda updated earlier, and exits.

Run it with the following command.

```bash
node test.js
```

When your code has several tests and you want a colorful concise summary, you can instead use the following.

```bash
node test.js | npx faucet
```

In reality, you'll want to create a random user at the beginning of each login test and remove it at the end. This will allow you to add multiple lambda tests while avoiding potential conflicts between test users and permissions.

> 

Reviewer: Would be great to have another unit test example where you are pulling attributes from the user object, performing logic on it, and then setting a value on a jwt. Maybe the user has a certain role ('admin', 'editor', 'viewer') and the JWT is given certain permissions based on those: (admin -> 'all', 'editor' -> 'read', 'write', viewer' -> 'read'). or something similar. But I think this kind of logic is where unit tests shine, plus showing how to mock up a user object would be useful.
